src/table.ts

```diff
@@ -866,7 +864,22 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
             rowKeys = null;
           }
         }
+
+        // If there was a row limit in the original request and
+        // we've already read all the rows, end the stream and
+        // do not retry.
+        if (hasLimit && rowsLimit === rowsRead) {
+          userStream.end();
+          return;
+        }
+        // If all the row keys and ranges are read, end the stream
+        // and do not retry.
+        if (!rowKeys && ranges.length === 0) {
+          userStream.end();
+          return;
+        }
       }
+
       if (rowKeys || ranges.length) {
         reqOpts.rows = {};
 
```

Comment:
Why do you need the if statements below, the loops will be no-ops for empty arrays